src/Uno.UI/UI/Xaml/Controls/TextBlock/SkiaTextLine.skia.cs

```diff
@@ -0,0 +1,70 @@
+﻿#nullable enable
+
+using SkiaSharp;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Windows.UI.Composition
+{
+	/// <summary>
+	/// Because Skia doesn't support fallback fonts for specific parts of text, we manually handle that in this class.
+	/// </summary>
+	internal class SkiaTextLine
+	{
+		private readonly List<(string Text, SKTypeface? Font)> _parts = new List<(string Text, SKTypeface? Font)>();
+		public IReadOnlyCollection<(string Text, SKTypeface? Font)> TextParts => _parts.AsReadOnly();
+
+		public SkiaTextLine(string text, SKTypeface? font, Func<string, SKFontStyleWeight, SKFontStyleWidth, SKFontStyleSlant, SKTypeface> getTypeFace)
+		{
+			if (text is null)
+			{
+				throw new ArgumentNullException(nameof(text));
+			}
+
+			if (text.Length == 0)
+			{
+				_parts.Add((text, font));
+				return;
+			}
+
+			var builder = new StringBuilder(capacity: text.Length);
+			builder.Append(text[0]);
+			SKTypeface? lastValidFont = GetFontForCharacter(text[0], font, getTypeFace);
+			for (var i = 1; i < text.Length; i++)
+			{
+				var character = text[i];
+				var currentFont = GetFontForCharacter(character, font, getTypeFace);
+				if (lastValidFont != currentFont)
+				{
+					_parts.Add((builder.ToString(), lastValidFont));
+					builder.Clear();
+					lastValidFont = currentFont;
+				}
+
+				builder.Append(character);
+			}
```

Comment:
> For most cases, we are likely going to stop after the first check. I'm interpreting your comment as if the second check is the expensive one, which won't happen for the majority of cases.

This loop is executed for every character, for every line and invokes `ContainsGlyph`, that's where I'm curious about the performance. I'm also wondering if your other font related PR fixes this without having to iterate manually over the characters and querying the font.

Reply:
This PR and the other PR are really different.

This one fixes characters that are not supported completely by the font. The other PR is when the characters are supported by the font, and they render, but in incorrect way (In Arabic for example, the exact same letter can have different shapes depending on where in a word it's. This is what the other PR for.

Regarding the performance, it's hard to tell without some way of benchmarking. Would some sort of caching solve this anyways (ie, a memoized func or something like that)?

Reply:
@jeromelaban Does it make sense to hide this under a feature flag? If yes, should the new behavior be disabled or enabled by default?

Also should I document the feature flag as experimental so we can remove it in future if needed?

Reply:
There's the text benchmark in the tests that should help to determine the performance impact: https://github.com/unoplatform/uno/blob/master/src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/TextBlockControl/TextBlock_MeasurePeformance.xaml

Reply:
@jeromelaban Can't see performance impact from the test you mentioned. But I'm not sure it actually hits this code path. It tests measure while the code path in question is used to render the text.

Anyway, I added a `var sw = Stopwatch.StartNew()` at the constructor beginning and logged few `$"{sw.ElapsedMilliseconds} - Text length: {text.Length}"`

The result is always zero. So there shouldn't be any performance impact here.

```
0 - Text length: 1
0 - Text length: 1
0 - Text length: 2
0 - Text length: 1
0 - Text length: 2
0 - Text length: 1
0 - Text length: 1
0 - Text length: 7
0 - Text length: 7
0 - Text length: 4
0 - Text length: 7
0 - Text length: 6
0 - Text length: 15
0 - Text length: 16
0 - Text length: 15
0 - Text length: 18
0 - Text length: 14
0 - Text length: 18
0 - Text length: 11
0 - Text length: 16
0 - Text length: 11
0 - Text length: 22
0 - Text length: 10
0 - Text length: 11
0 - Text length: 12
0 - Text length: 13
0 - Text length: 15
0 - Text length: 13
0 - Text length: 15
0 - Text length: 21
0 - Text length: 31
0 - Text length: 17
0 - Text length: 20
0 - Text length: 16
0 - Text length: 10
0 - Text length: 17
0 - Text length: 14
0 - Text length: 1
0 - Text length: 1
0 - Text length: 27
0 - Text length: 19
0 - Text length: 21
0 - Text length: 15
0 - Text length: 18
0 - Text length: 13
0 - Text length: 12
0 - Text length: 12
0 - Text length: 11
0 - Text length: 9
0 - Text length: 20
0 - Text length: 12
0 - Text length: 17
0 - Text length: 23
0 - Text length: 11
0 - Text length: 17
0 - Text length: 21
0 - Text length: 15
0 - Text length: 10
0 - Text length: 26
0 - Text length: 13
0 - Text length: 11
0 - Text length: 17
0 - Text length: 12
0 - Text length: 13
0 - Text length: 16
0 - Text length: 10
0 - Text length: 23
0 - Text length: 7
0 - Text length: 23
0 - Text length: 22
0 - Text length: 11
0 - Text length: 14
0 - Text length: 21
0 - Text length: 8
0 - Text length: 19
0 - Text length: 13
0 - Text length: 8
0 - Text length: 16
0 - Text length: 17
0 - Text length: 21
0 - Text length: 15
0 - Text length: 15
0 - Text length: 13
0 - Text length: 9
0 - Text length: 14
0 - Text length: 12
0 - Text length: 16
0 - Text length: 20
0 - Text length: 17
0 - Text length: 22
0 - Text length: 16
0 - Text length: 18
0 - Text length: 17
0 - Text length: 13
0 - Text length: 17
0 - Text length: 11
0 - Text length: 10
0 - Text length: 15
0 - Text length: 13
0 - Text length: 18
0 - Text length: 9
0 - Text length: 16
0 - Text length: 11
0 - Text length: 27
0 - Text length: 19
0 - Text length: 21
0 - Text length: 12
0 - Text length: 16
0 - Text length: 14
0 - Text length: 15
0 - Text length: 14
0 - Text length: 16
0 - Text length: 11
0 - Text length: 11
0 - Text length: 9
0 - Text length: 19
0 - Text length: 14
0 - Text length: 12
0 - Text length: 15
0 - Text length: 14
0 - Text length: 28
0 - Text length: 14
0 - Text length: 11
0 - Text length: 17
0 - Text length: 12
0 - Text length: 28
0 - Text length: 20
0 - Text length: 25
0 - Text length: 4
0 - Text length: 16
0 - Text length: 33
0 - Text length: 22
0 - Text length: 32
0 - Text length: 28
0 - Text length: 41
0 - Text length: 41
0 - Text length: 18
0 - Text length: 21
0 - Text length: 40
0 - Text length: 33
0 - Text length: 40
0 - Text length: 41
0 - Text length: 11
0 - Text length: 28
0 - Text length: 39
0 - Text length: 39
0 - Text length: 23
0 - Text length: 23
0 - Text length: 44
0 - Text length: 33
0 - Text length: 42
0 - Text length: 23
0 - Text length: 24
0 - Text length: 28
0 - Text length: 26
0 - Text length: 1
0 - Text length: 1
0 - Text length: 33
0 - Text length: 11
0 - Text length: 58
0 - Text length: 18
0 - Text length: 11
0 - Text length: 22
0 - Text length: 11
0 - Text length: 18
0 - Text length: 14
0 - Text length: 11
0 - Text length: 32
0 - Text length: 11
0 - Text length: 14
0 - Text length: 40
0 - Text length: 31
0 - Text length: 11
0 - Text length: 41
0 - Text length: 11
0 - Text length: 15
0 - Text length: 15
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 2
0 - Text length: 19
0 - Text length: 43
0 - Text length: 21
0 - Text length: 15
0 - Text length: 11
0 - Text length: 18
0 - Text length: 11
0 - Text length: 19
0 - Text length: 43
0 - Text length: 1
0 - Text length: 21
0 - Text length: 15
0 - Text length: 43
0 - Text length: 43
0 - Text length: 43
0 - Text length: 43
0 - Text length: 43
0 - Text length: 43
0 - Text length: 43
0 - Text length: 43
0 - Text length: 43
0 - Text length: 11
0 - Text length: 33
0 - Text length: 11
0 - Text length: 43
0 - Text length: 43
0 - Text length: 43
0 - Text length: 43
0 - Text length: 43
0 - Text length: 43
0 - Text length: 43
0 - Text length: 43
0 - Text length: 43
0 - Text length: 32
0 - Text length: 21
0 - Text length: 39
0 - Text length: 20
0 - Text length: 28
0 - Text length: 36
0 - Text length: 33
0 - Text length: 19
0 - Text length: 25
0 - Text length: 20
0 - Text length: 20
0 - Text length: 13
0 - Text length: 15
0 - Text length: 119
0 - Text length: 129
0 - Text length: 147
0 - Text length: 137
0 - Text length: 11
0 - Text length: 78
0 - Text length: 33
0 - Text length: 11
0 - Text length: 78
0 - Text length: 20
0 - Text length: 20
0 - Text length: 13
0 - Text length: 15
0 - Text length: 119
0 - Text length: 129
0 - Text length: 147
0 - Text length: 137
0 - Text length: 33
0 - Text length: 32
0 - Text length: 23
0 - Text length: 16
0 - Text length: 25
0 - Text length: 28
0 - Text length: 30
0 - Text length: 37
0 - Text length: 34
0 - Text length: 33
0 - Text length: 7
0 - Text length: 10
0 - Text length: 5
0 - Text length: 9
0 - Text length: 4
0 - Text length: 8
0 - Text length: 6
0 - Text length: 6
0 - Text length: 5
0 - Text length: 9
0 - Text length: 10
0 - Text length: 4
0 - Text length: 7
0 - Text length: 10
0 - Text length: 5
0 - Text length: 9
0 - Text length: 4
0 - Text length: 8
0 - Text length: 6
0 - Text length: 6
0 - Text length: 5
0 - Text length: 9
0 - Text length: 10
0 - Text length: 4
0 - Text length: 11
0 - Text length: 20
0 - Text length: 11
0 - Text length: 7
0 - Text length: 10
0 - Text length: 5
0 - Text length: 9
0 - Text length: 4
0 - Text length: 8
0 - Text length: 6
0 - Text length: 6
0 - Text length: 5
0 - Text length: 9
0 - Text length: 10
0 - Text length: 4
0 - Text length: 7
0 - Text length: 10
0 - Text length: 5
0 - Text length: 9
0 - Text length: 4
0 - Text length: 8
0 - Text length: 6
0 - Text length: 6
0 - Text length: 5
0 - Text length: 9
0 - Text length: 10
0 - Text length: 4
0 - Text length: 33
0 - Text length: 11
0 - Text length: 30
0 - Text length: 4
0 - Text length: 4
0 - Text length: 4
0 - Text length: 4
0 - Text length: 4
0 - Text length: 4
0 - Text length: 4
0 - Text length: 4
0 - Text length: 4
0 - Text length: 22
0 - Text length: 37
0 - Text length: 46
0 - Text length: 17
0 - Text length: 26
0 - Text length: 25
0 - Text length: 14
0 - Text length: 27
0 - Text length: 27
0 - Text length: 23
0 - Text length: 22
0 - Text length: 48
0 - Text length: 29
0 - Text length: 4
0 - Text length: 7
0 - Text length: 2
0 - Text length: 1
0 - Text length: 9
0 - Text length: 2
0 - Text length: 1
0 - Text length: 11
0 - Text length: 6
0 - Text length: 1
0 - Text length: 17
0 - Text length: 1
0 - Text length: 1
0 - Text length: 5
0 - Text length: 6
0 - Text length: 1
0 - Text length: 11
0 - Text length: 22
0 - Text length: 11
0 - Text length: 29
0 - Text length: 4
0 - Text length: 1
0 - Text length: 7
0 - Text length: 2
0 - Text length: 1
0 - Text length: 9
0 - Text length: 2
0 - Text length: 1
0 - Text length: 11
0 - Text length: 6
0 - Text length: 1
0 - Text length: 17
0 - Text length: 1
0 - Text length: 1
0 - Text length: 5
0 - Text length: 6
0 - Text length: 38
0 - Text length: 27
0 - Text length: 28
0 - Text length: 34
0 - Text length: 46
0 - Text length: 17
0 - Text length: 29
0 - Text length: 21
0 - Text length: 29
0 - Text length: 19
0 - Text length: 14
0 - Text length: 13
0 - Text length: 11
0 - Text length: 39
0 - Text length: 27
0 - Text length: 11
0 - Text length: 39
0 - Text length: 14
0 - Text length: 5
0 - Text length: 7
0 - Text length: 5
0 - Text length: 11
0 - Text length: 17
0 - Text length: 11
0 - Text length: 5

```

Reply:
Thanks, let's try this out then.